platform/linux-generic/pktio/socket_mmap.c

```diff
@@ -243,7 +247,21 @@ static inline unsigned pkt_mmap_v2_rx(pktio_entry_t *pktio_entry,
 			memmove(mac, mac + vlan_len, 2 * _ODP_ETHADDR_LEN);
 			type  = (uint16_t *)(uintptr_t)
 				(mac + 2 * _ODP_ETHADDR_LEN);
+
+			#ifdef TP_STATUS_VLAN_TPID_VALID
 			*type = odp_cpu_to_be_16(tp_hdr->tp_vlan_tpid);
+			#else
+			/* Fallback for old kernels (< v3.14) */
+			uint16_t *type2;
+
+			type2 = (uint16_t *)(uintptr_t)(mac + (2 * _ODP_ETHADDR_LEN) + vlan_len);
+			/* Recreate TPID 0x88a8 for double tagged and 0x8100 for single tagged */
+			if (*type2 == odp_cpu_to_be_16(0x8100))
+				*type = odp_cpu_to_be_16(0x88a8);
```

Comment:
This seems to be just an educated guess. Isn't there any better way? Sometimes TPID values 0x8100 and even 0x9100 are used in the outer VLAN tag with double tagging,

Reply:
The necessary header information is already lost at this point. I added a debug print about this.